tools/toolchains/arm.py

```diff
@@ -546,12 +580,21 @@ def __init__(self, target, *args, **kwargs):
         self.ar = join(TOOLCHAIN_PATHS["ARMC6"], "armar")
         self.elf2bin = join(TOOLCHAIN_PATHS["ARMC6"], "fromelf")
 
+        # Adding this for safety since this inherits the `version_check` function
+        # but does not call the constructor of ARM_STD, so the `product_name` variable
+        # is not initialized.
+        self.product_name = None
+
     def _get_toolchain_labels(self):
         if getattr(self.target, "default_toolchain", "ARM") == "uARM":
             return ["ARM", "ARM_MICRO", "ARMC6"]
         else:
             return ["ARM", "ARM_STD", "ARMC6"]
 
+    @property
+    def is_mbed_studio_armc6(self):
+        return self.product_name and "Mbed Studio" in self.product_name
```

Comment:
I'm not convinced that mbed os tools should have special logic based on a single product. Not sure if there is a better way of fixing it but it should be worth double checking before merging it.

Reply:
It also limit options to make any changes in product name if needed. As old mbed os versions will be already released.

Reply:
> I'm not convinced that mbed os tools should have special logic based on a single product. 

Considering the compiler has special behavior in this single product, I don't know if we have much of a choice 😄. The license shipped with Mbed Studio **must** be supplied to the compiler shipped with Mbed Studio in order to build anything. However if you have any alternatives/suggestions do let me know.

> It also limit options to make any changes in product name if needed. As old mbed os versions will be already released.

This is the only way I know of to identify this build of the compiler. As you can see, the output of `armclang --vsn` is quite similar between the different builds:

ARMC6 Professional:
```
Product: ARM Compiler 6.11 Professional
Component: ARM Compiler 6.11
Tool: armclang [5d3b4200]
```

ARMC6 from Mbed Studio:
```
<removed license error for brevity>
Product: ARM Compiler 6.11 for Mbed Studio
Component: ARM Compiler 6.11
Tool: armclang [5d3b3c00]
```

And we shouldn't be switching this behavoir based on the commit since that will change with new releases of the compiler.

Reply:
I agree with @arekzaluski, managing specific cases like this is problematic and prone to breaking. For example, we plan to move the installation directory of the tools in the next release of Mbed Studio.
At a higher level, requiring Mbed Studio to be installed for this functionality to work makes little sense IMO.

Reply:
@thegecko I understand your concern, it definitely reduces the flexibility of how Mbed Studio is internally organized.

> For example, we plan to move the installation directory of the tools in the next release of Mbed Studio.

This isn't immediately a problem for this implementation, as it only requires the `ac6-license.dat` file to be placed in any parent directory of the compiler binary.

> At a higher level, requiring Mbed Studio to be installed for this functionality to work makes little sense IMO.

Mbed Studio is only required to be installed if you wish to take advantage of the development license and corresponding compiler provided by Mbed Studio. You can of course still use the professional versions of the Arm Compiler 6 and license.

---

Another option is that I completely remove all of the "license detection" features at the expense of user experience. This would mean the user would now be responsible for the following:

1. Find the `ac6-license.dat` file in the Mbed Studio installation
1. Add the `ARMLMD_LICENSE_FILE` environment variable to their shell and set it to the path to the `ac6-license.dat` file.

We can of course document all of this, but that also means we need to keep this documentation up to date whenever the structure of Mbed Studio changes.

If the license detection feature is something that we need to ship though, I really need an alternative suggestion from your team. I'll happily parse a file to get the exact location. Maybe you could place a dedicated file next to the armc6 binaries? That file could point to the correct location of the license.

Reply:
I was thinking about it and I think that the best way would be to remove license detection feature from mbed-os tools. Instead, it should be `applications` (Mbed Studio, mbed-cli, possible other in the future) that controls specific compilation behavior. Only they know location of all of the files.

At the moment Mbed Studio handles cases when `ARMLMD_LICENSE_FILE` env variable is set or/and AC6 compiler is installed on user's system. It simply overrides it and set its own path for both license and compiler location. It overrides it only for a single build process so user's environment variables are not changed.

I believe that the best way forward is to add similar logic in mbed-cli. For example installers of mbed-cli can be shipped with ARMC6 and valid license. Installer knows installation location so it can ask user at the end of installation if env variables should be updated or not.
Of course it should be also well documented for users that do not use installers to install mbed-cli.

@bridadan What do you think about it?

Reply:
Seems like the least contentious way forward, so I'll go with that. Moving this functionality to the installers doesn't sound like a bad idea either 👍

Reply:
@bridadan Checking since this PR has passed CI checks.

I'm assuming the above will be planned for 5.13, and _not_ squeezed into 5.12.0, correct?

Reply:
Based on the feedback from the Mbed Studio folks it sounds like they are not comfortable with this PR going in as-is for 5.12, so I think I have to make the tool changes now. The installers will not be ready though for sometime.

Reply:
For what its worth I have the changes almost ready. I don't have access to a mac to test though.